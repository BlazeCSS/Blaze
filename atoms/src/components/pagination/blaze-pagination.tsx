import { Component, Event, EventEmitter, Method, Prop, State } from '@stencil/core';

@Component({
  tag: 'blaze-pagination'
})
export class Pagination {

  @Prop() page: number = 1;
  @Prop() pages: number = 1;
  @State() _currentPage: number;
  @Event() onPageChange: EventEmitter;

  @Method()
  goToPage(page: number) {
    if (page > 0 && page <= this.pages) {
      this._currentPage = page;
      this.onPageChange.emit(this._currentPage);
    }
  }

  @Method()
  currentPage() {
    return this._currentPage;
  }

  componentWillLoad() {
    this._currentPage = this.page;
  }

  render() {
    return (
      <div class="c-pagination">
        <div class="c-pagination__controls c-pagination__controls--backward">
          <button class="c-pagination__control" onClick={() => this.goToPage(1)} disabled={this._currentPage === 1}>&laquo;</button>
          <button class="c-pagination__control" onClick={() => this.goToPage(this._currentPage - 1)} disabled={this._currentPage === 1}>&lsaquo;</button>
        </div>
        <div class="c-pagination__controls">
          {this._currentPage > 1 && <span class="c-pagination__ellipsis">&hellip;</span>}
          {this._currentPage > 1 && <button onClick={() => this.goToPage(this._currentPage - 1)} class="c-pagination__page">{this._currentPage - 1}</button>}
          <button class="c-pagination__page c-pagination__page--current">{this._currentPage}</button>
          {this._currentPage < this.pages && <button onClick={() => this.goToPage(this._currentPage + 1)} class="c-pagination__page">{this._currentPage + 1}</button>}
          {this._currentPage < this.pages && <span class="c-pagination__ellipsis">&hellip;</span>}
        </div>
        <div class="c-pagination__controls c-pagination__controls--forward">
          <button class="c-pagination__control" onClick={() => this.goToPage(this._currentPage + 1)} disabled={this._currentPage === this.pages}>&rsaquo;</button>
          <button class="c-pagination__control" onClick={() => this.goToPage(this.pages)} disabled={this._currentPage === this.pages}>&raquo;</button>
        </div>
      </div>
    );
  }
}
