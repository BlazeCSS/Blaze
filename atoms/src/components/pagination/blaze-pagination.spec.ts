import { flush, render } from '@stencil/core/testing';
import { Pagination } from './blaze-pagination';

describe('Pagination', () => {
  it('should build', () => {
    expect(new Pagination()).toBeTruthy();
  });

  let element;

  const snapIt = (name, html) => {
    it(name, async () => {
      element = await render({
        components: [Pagination],
        html
      });
      await flush(element);

      expect(element).toMatchSnapshot();
    });
  };

  snapIt(
    'should work without parameters',
    '<blaze-pagination></blaze-pagination>'
  );

  snapIt(
    'renders the correct buttons when pages set',
    '<blaze-pagination pages="15"></blaze-pagination>'
  );

  it('has a go to page function', async () => {
    element = await render({
      components: [Pagination],
      html: '<blaze-pagination pages="15"></blaze-pagination>'
    });
    await flush(element);
    expect(element).toMatchSnapshot();

    element.goToPage(4);
    await flush(element);

    expect(element).toMatchSnapshot();
  });

  it('emits a page changed event', async () => {
    element = await render({
      components: [Pagination],
      html: '<blaze-pagination pages="15"></blaze-pagination>'
    });
    await flush(element);

    element.addEventListener('onPageChange', (e) => {
      expect(e.detail).toEqual(4);
    });
    element.goToPage(4);
  });

});
