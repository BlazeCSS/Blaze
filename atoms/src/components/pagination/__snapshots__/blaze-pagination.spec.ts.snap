// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination has a go to page function 1`] = `
<blaze-pagination
  class="hydrated"
  pages="15"
>
  <!---->
  <div
    class="c-pagination"
  >
    <div
      class="c-pagination__controls c-pagination__controls--backward"
    >
      <button
        class="c-pagination__control"
        disabled=""
      >
        «
      </button>
      <button
        class="c-pagination__control"
        disabled=""
      >
        ‹
      </button>
    </div>
    <div
      class="c-pagination__controls"
    >
      
      <button
        class="c-pagination__page c-pagination__page--current"
      >
        1
      </button>
      <button
        class="c-pagination__page"
      >
        2
      </button>
      <span
        class="c-pagination__ellipsis"
      >
        …
      </span>
    </div>
    <div
      class="c-pagination__controls c-pagination__controls--forward"
    >
      <button
        class="c-pagination__control"
      >
        ›
      </button>
      <button
        class="c-pagination__control"
      >
        »
      </button>
    </div>
  </div>
</blaze-pagination>
`;

exports[`Pagination has a go to page function 2`] = `
<blaze-pagination
  class="hydrated"
  pages="15"
>
  <!---->
  <div
    class="c-pagination"
  >
    <div
      class="c-pagination__controls c-pagination__controls--backward"
    >
      <button
        class="c-pagination__control"
      >
        «
      </button>
      <button
        class="c-pagination__control"
      >
        ‹
      </button>
    </div>
    <div
      class="c-pagination__controls"
    >
      <span
        class="c-pagination__ellipsis"
      >
        …
      </span>
      <button
        class="c-pagination__page"
      >
        3
      </button>
      <button
        class="c-pagination__page c-pagination__page--current"
      >
        4
      </button>
      <button
        class="c-pagination__page"
      >
        5
      </button>
      <span
        class="c-pagination__ellipsis"
      >
        …
      </span>
    </div>
    <div
      class="c-pagination__controls c-pagination__controls--forward"
    >
      <button
        class="c-pagination__control"
      >
        ›
      </button>
      <button
        class="c-pagination__control"
      >
        »
      </button>
    </div>
  </div>
</blaze-pagination>
`;

exports[`Pagination renders the correct buttons when pages set 1`] = `
<blaze-pagination
  class="hydrated"
  pages="15"
>
  <!---->
  <div
    class="c-pagination"
  >
    <div
      class="c-pagination__controls c-pagination__controls--backward"
    >
      <button
        class="c-pagination__control"
        disabled=""
      >
        «
      </button>
      <button
        class="c-pagination__control"
        disabled=""
      >
        ‹
      </button>
    </div>
    <div
      class="c-pagination__controls"
    >
      
      <button
        class="c-pagination__page c-pagination__page--current"
      >
        1
      </button>
      <button
        class="c-pagination__page"
      >
        2
      </button>
      <span
        class="c-pagination__ellipsis"
      >
        …
      </span>
    </div>
    <div
      class="c-pagination__controls c-pagination__controls--forward"
    >
      <button
        class="c-pagination__control"
      >
        ›
      </button>
      <button
        class="c-pagination__control"
      >
        »
      </button>
    </div>
  </div>
</blaze-pagination>
`;

exports[`Pagination should work without parameters 1`] = `
<blaze-pagination
  class="hydrated"
>
  <!---->
  <div
    class="c-pagination"
  >
    <div
      class="c-pagination__controls c-pagination__controls--backward"
    >
      <button
        class="c-pagination__control"
        disabled=""
      >
        «
      </button>
      <button
        class="c-pagination__control"
        disabled=""
      >
        ‹
      </button>
    </div>
    <div
      class="c-pagination__controls"
    >
      
      <button
        class="c-pagination__page c-pagination__page--current"
      >
        1
      </button>
      
    </div>
    <div
      class="c-pagination__controls c-pagination__controls--forward"
    >
      <button
        class="c-pagination__control"
        disabled=""
      >
        ›
      </button>
      <button
        class="c-pagination__control"
        disabled=""
      >
        »
      </button>
    </div>
  </div>
</blaze-pagination>
`;
